Enable Prometheus metrics on Ceph mgr

galexrt opened the pull request:
Fixes #1111.

***

This **a)** create a service in front of the Ceph mgrs and **b)** enable the mgrs Prometheus module to provide prometheus metrics (instead of the rook-api).

This hasn't been tested yet. Will do in the next few days.

**ToDo**:
- [x] Update prometheus-operator service monitor manifest
- [x] Test it

/cc @jpds

galexrt commented:
Building the rook image, running just `make` in the `images/` directory, results in the following:
**Without the dependencies**:
```
build-da9b3fc0/toolbox-amd64                                                     latest               eb010b560aa5        About an hour ago   419MB
build-da9b3fc0/toolbox-base-amd64                                                latest               cb1f8737cbf1        About an hour ago   403MB
build-da9b3fc0/rook-amd64                                                        latest               1a978894c4b7        About an hour ago   347MB
build-da9b3fc0/rook-base-amd64                                                   latest               8431ef20591c        About an hour ago   293MB
```

**With the new python dependencies**:
```
build-da9b3fc0/toolbox-amd64                                                     latest               46428021169f        22 minutes ago      462MB
build-da9b3fc0/toolbox-base-amd64                                                latest               3acd76ff79b1        22 minutes ago      445MB
build-da9b3fc0/rook-amd64                                                        latest               1d019205b43f        23 minutes ago      389MB
build-da9b3fc0/rook-base-amd64                                                   latest               b83ad9c03671        23 minutes ago      335MB
```

jbw976 commented on pkg/operator/mgr/mgr.go:
```diff
@@ -89,6 +104,31 @@ func (c *Cluster) Start() error {
+		Type:     v1.ServiceTypeClusterIP,
+		Ports: []v1.ServicePort{
+			{
+				Name:     "mgr",
```
who do you imagine will be a consumer of the service port for `mgr` itself?  I imagine that the service port for the metrics will be consumed by external scrapers (prometheus), but i'm not sure who would need a service port for `mgr`.

galexrt replied:
Right, I remove the `mgr` port.

jbw976 commented on pkg/operator/mgr/mgr.go:
```diff
@@ -187,6 +239,14 @@ func (c *Cluster) createKeyring(clusterName, name string) error {
 }
 
+// Ceph docs about the prometheus module: http://docs.ceph.com/docs/master/mgr/prometheus/
+func (c *Cluster) enablePrometheusModule(clusterName string) error {
```
the metrics documentation/example should be updated to point at this new ceph-mgr module endpoint instead.

https://github.com/rook/rook/blob/master/Documentation/k8s-monitoring.md
https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/monitoring/service-monitor.yaml#L14

galexrt replied:
The docs don't mention the API nor the Ceph mgrs right now, should that be added?

jbw976 commented on pkg/operator/mgr/mgr.go:
```diff
@@ -89,6 +104,31 @@ func (c *Cluster) Start() error {
 	return nil
 }
 
+func (c *Cluster) makeService(name string) *v1.Service {
```
will our existing teardown/cleanup instructions cover this dynamically created resource as well or do those need to be updated?

galexrt replied:
@jbw976 The `kubernetes.md` teardown/cleanup should suffice because it deletes the namespace. Is that good or should a line for deleting the service be added?

jbw976 commented on pkg/operator/mgr/mgr.go:
```diff
@@ -187,6 +239,14 @@ func (c *Cluster) createKeyring(clusterName, name string) error {
 
+// Ceph docs about the prometheus module: http://docs.ceph.com/docs/master/mgr/prometheus/
+func (c *Cluster) enablePrometheusModule(clusterName string) error {
+	if err := client.MgrEnableModule(c.context, clusterName, prometheusModuleName); err != nil {
```
do you know much about the format of the metrics exported by this module as compared to what we built our solution on from ceph-exporter?
https://github.com/digitalocean/ceph_exporter

There is a nice grafana dashboard that is built around those that folks are using.  Do you know if the metrics from the ceph-mgr module are in the same format?
https://grafana.com/dashboards/917

hunter replied:
I think there is a new official metrics grafana dashboard that was announced at the Openstack Summit this week. The json can be found: https://github.com/ceph/cephmetrics/tree/wip-prometheus/dashboards/prometheus

galexrt replied:
@hunter I have tried to use the dashboard from the `cephmetrics` repo, but they aren't working..
Only getting empty dashboards out of the json files.

dangula commented:
retest this please

galexrt commented:
@jbw976 I tested it locally in minikube and @jpds will also test it soon.
The metrics output looked like this:
![2017-11-26-135854_1920x1080_scrot2](https://user-images.githubusercontent.com/3718398/33240166-18e304e6-d2b2-11e7-9094-5b431cbc41cb.png)

PTAL

jpds commented:
I got this working correctly with @galexrt's image and it looked good to me (save the inactive mgr1 endpoint).

The fact that it's `down` could be an issue as I have a prometheus alert configured to detect these. 😄

Perhaps it's best to have an mgr SVC per mgr?

jbw976 commented:
I'd like to get this work merged into master because I'm also thinking about getting the ceph-mgr restful module enabled and that would use some of this code.  are there other remaining issues for this PR?
1. grafana dashboards
1. standby ceph-mgr instances are reported as down

kokhang commented:
Since we are planning to drop the rook-api, this would be needed. @galexrt Any update on this?

galexrt commented:
@kokhang As @jbw976 wrote one mgr in the current Ceph version is "down" due it being in standby mode.
There are plans from Ceph to allow a standby mgr to proxy to the current active mgr (making it not "down"), but this is still a todo from what I know.
The second issue I'm raising is that at least I couldn't get the dashboards @hunter mentioned to work in Grafana (I'm not sure if they are even Grafana dashboards or are Prometheus "Consoles" dashboards).

I think the bigger part here is to solve the first issue with the standby mgr being down (connection refused). Any good ideas how we could solve this properly?

galexrt commented:
There are two tasks coming after this:
* Document the "breaking" change in the Changelog
* Create a Grafana dashboards

jcsp commented:
FYI the ability to redirect/proxy via standby mgr nodes is in ceph's master (and luminous) branches.  The exact behaviour that the prometheus module should have is still a subject of discussion (ping @jan--f)

galexrt commented:
@jbw976 Is there anything that should be changed?
I will test it with Ceph 12.2.2 version in minikube again to see if everything works. @jbw976 I will send you an example of the metrics exposed.

jbw976 commented:
Thanks for the update @galexrt, let me know when you've tested it with minikube.

Should we add a pointer in our docs to the Prometheus dashboard(s) that visualize these new metrics?

jbw976 commented:
Also, do you understand the upgrade issue you were referring to in the call on Monday?  Is there a workaround for that? does it repro 100%? do we need docs or breaking change notice for that?

galexrt commented:
I couldn't reproduce the failure of Mons when telling the Mgr to enable the Prometheus module in Minikube anymore.
Still I think it would be good to add a note to the changelog that it is probably safer to update in the following order:
1. Rook Ceph (first Mgr, then Mons and the other components like OSD, API, RGW, etc)
2. Rook Operator (then creates the Mgr service and enables the Prometheus module on the Mons

Side note: When reverting the commit to set replicas to `2` for Mgrs we didn't mention that the users should set the replicas to `1` (as in the code now). If the user doesn't do that he may receive an alert upon updating/creating the `ServiceMonitor` because one of the Mgrs is down because of that.

jbw976 commented:
Let me know if all these loose ends are wrapped up:
1. It would be nice add a pointer in our docs to the Prometheus dashboard(s) that visualize these new metrics, somewhere in https://rook.io/docs/rook/master/k8s-monitoring.html
1. Could you go ahead and add the note about upgrade order to the pending release notes?
1. Could your share your gist of the metrics output here?

If i understand your side note, I believe I'm OK with that.  The default path is for just 1 ceph-mgr instance, so I don't think we need to document anything here.

galexrt commented:
1. I will add a link to the github.com/ceph/cephmetrics dashboards directory, they are Prometheus and **not** Grafana dashboards. I'm going to build Grafana dashboards myself ASAP when this PR is merged and publish them to rook repo and/or Grafana.com.
2. Will do.
3. The gist with the metrics can be found here: https://gist.github.com/galexrt/b0bb2bc484163a1626f4d73c05f8c2af.

jan--f commented:
I'm working on a patch for the prometheus module, that adds a StanbyModule. This means that prometheus will simply be able to scrape all mgr daemons. The active daemon will return the metrics, all passive daemons will return an empty answer.

@galexrt In [DeepSea](https://github.com/SUSE/DeepSea) we already have a bunch of grafana dashboards. I'd be happy to collaborate on a somewhat unified set of dashboards.

galexrt commented:
@jan--f Nice, that is good to hear. Please let us know when the patch is merged.
When @jan--f patch for the module is merged we can go and revert the revert of using two mgrs. ;)

@jbw976 I have added this change to the pending release notes.

jan--f commented:
prometheus module patch is here https://github.com/ceph/ceph/pull/19744

galexrt commented:
@jan--f I will get in touch with you about the dashboards once I have the Ceph MGR monitoring running in one of my clusters.

@jan--f Do you have an ETA for the patch you have submitted?
@jbw976 Depending on how long it might take for the patch to the Ceph Prometheus metrics module to be available to us in the currently used Ceph release.
I think we should go ahead and get this merged and then with the PR that reverts the revert of running only one mgr, get the patched version of Prometheus metrics module patch in Rook. What do you think?

jbw976 commented on PendingReleaseNotes.md:
```diff
@@ -1,17 +1,21 @@
 # Major Themes
 
 ## Action Required
+- The order in which the Pods should be updated is: MON, MGR, MDS, OSD, Rook API and then the Rook Operator.
```
this comment says that the MGRs need to be first with this monitoring change: https://github.com/rook/rook/pull/1154#issuecomment-354342886

which is correct?  if you can't repro the crash during upgrade, maybe it's not worth mentioning it here actually.  It could have been a bug that was fixed with 12.2.2?

galexrt replied:
Could be, I removed it.

galexrt commented:
@jbw976 I added the missing link to the dashboards, I also mentioned the Grafana ones from @jan--f.
The `PendingReleaseNotes.md` has also changed accordingly PTAL
